src/v/archival/service.cc

```diff
@@ -0,0 +1,378 @@
+      config::shard_local_cfg().archival_storage_s3_access_key,
+      "archival_storage_s3_access_key"));
+    auto region = s3::aws_region_name(get_value_or_throw(
+      config::shard_local_cfg().archival_storage_s3_region,
+      "archival_storage_s3_region"));
+    auto s3_conf = co_await s3::configuration::make_configuration(
+      access_key, secret_key, region);
+    archival::configuration cfg{
+      .client_config = std::move(s3_conf),
+      .bucket_name = s3::bucket_name(get_value_or_throw(
+        config::shard_local_cfg().archival_storage_s3_bucket,
+        "archival_storage_s3_bucket")),
+      .upload_policy = {}, // The values are ignored because only one policy
+      .delete_policy{},    // is implemented for now.
+      .interval
+      = config::shard_local_cfg().archival_storage_upload_interval.value(),
+      .gc_interval
+      = config::shard_local_cfg().archival_storage_gc_interval.value(),
+      .connection_limit = s3_connection_limit(
+        config::shard_local_cfg().archival_storage_max_connections.value())};
+    co_return std::move(cfg);
+}
+
+scheduler_service_impl::scheduler_service_impl(
+  const configuration& conf,
+  ss::sharded<storage::api>& api,
+  ss::sharded<cluster::partition_manager>& pm)
+  : _conf(conf)
+  , _partition_manager(pm)
+  , _storage_api(api)
+  , _jitter(conf.interval, 1ms)
+  , _gc_jitter(conf.gc_interval, 1ms)
+  , _conn_limit(conf.connection_limit()) {}
+
+void scheduler_service_impl::rearm_timer() {
+    (void)ss::with_gate(_gate, [this] {
+        auto next = _jitter();
+        return workflow()
+          .finally([this, next] {
+              if (_gate.is_closed()) {
+                  return;
+              }
+              _timer.rearm(next);
+          })
+          .handle_exception([](std::exception_ptr e) {
+              vlog(archival_log.info, "Error in timer callback: {}", e);
+          });
+    });
+}
+void scheduler_service_impl::start() {
+    _timer.set_callback([this] { rearm_timer(); });
+    _timer.rearm(_jitter());
+}
+
+ss::future<> scheduler_service_impl::stop() {
+    vlog(archival_log.info, "Scheduler service stop");
+    _timer.cancel();
+    return _gate.close();
+}
+
+ss::future<> scheduler_service_impl::workflow() {
+    static thread_local size_t niter = 0;
+    vlog(archival_log.trace, "Start S3 maintanance cycle #{}", niter);
+    co_await reconcile_archivers();
+    co_await run_uploads();
+    co_await run_cleanup();
+    vlog(archival_log.trace, "Complete S3 maintanance cycle #{}", niter);
+    niter++;
+}
+
+ss::lw_shared_ptr<ntp_archiver> scheduler_service_impl::get_upload_candidate() {
+    return _queue.get_upload_candidate();
+}
+
+ss::lw_shared_ptr<ntp_archiver> scheduler_service_impl::get_delete_candidate() {
+    return _queue.get_upload_candidate();
+}
+
+ss::future<>
+scheduler_service_impl::create_archivers(std::vector<model::ntp> to_create) {
+    return ss::parallel_for_each(to_create, [this](const model::ntp& ntp) {
```

Comment:
Keep in mind the lifetime of `to_create`. I usually move collections into a ss::do_with() before I iterate over them with `ss::parallel_for_each` or `ss::do_for_each`.